iphone/Classes/KrollContext.m

```diff
@@ -1610,36 +1622,54 @@ - (void)dealloc
 
 - (void)invalidateAllTimers
 {
-  for (NSNumber *timerIdentifier in self.timers.allKeys) {
+  for (NSNumber *timerIdentifier in self.timers) {
     [self clearIntervalWithIdentifier:timerIdentifier.unsignedIntegerValue];
   }
 }
 
-- (NSUInteger)setIntervalFromArguments:(NSArray<JSValue *> *)arguments shouldRepeat:(BOOL)shouldRepeat
+- (NSUInteger)setInterval:(double)interval withCallback:(JSValue *)callback shouldRepeat:(BOOL)shouldRepeat
 {
-  NSMutableArray *callbackArgs = [arguments.mutableCopy autorelease];
-  JSValue *callbackFunction = [callbackArgs objectAtIndex:0];
-  [callbackArgs removeObjectAtIndex:0];
-  double interval = [[callbackArgs objectAtIndex:0] toDouble] / 1000;
-  [callbackArgs removeObjectAtIndex:0];
+  // interval is optional, should default to 0 per spec, but let's enforce at least 1 ms minimum (like Node)
+  if (isnan(interval) || interval < 1) {
+    interval = 1; // defaults to 1ms
+  }
+  interval = interval / 1000.0; // convert from ms to seconds
+
+  // Handle additional arguments being passed in
+  NSMutableArray<JSValue *> *callbackArgs = [[JSContext currentArguments].mutableCopy autorelease];
+  if ([callbackArgs count] > 1) {
+    [callbackArgs removeObjectAtIndex:1]; // drop interval
+  }
   NSNumber *timerIdentifier = @(self.nextTimerIdentifier++);
-  NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:interval
-                                                   repeats:shouldRepeat
-                                                     block:^(NSTimer *_Nonnull timer) {
-                                                       [callbackFunction callWithArguments:callbackArgs];
-                                                       if (!shouldRepeat) {
-                                                         [self.timers removeObjectForKey:timerIdentifier];
-                                                       }
-                                                     }];
-  self.timers[timerIdentifier] = timer;
+  NSTimer *timer = [NSTimer timerWithTimeInterval:interval target:self selector:@selector(callJsCallback:) userInfo:callbackArgs repeats:shouldRepeat];
+  [[NSRunLoop mainRunLoop] addTimer:timer forMode:NSRunLoopCommonModes];
+
+  [self.timers setObject:timer forKey:timerIdentifier];
   return [timerIdentifier unsignedIntegerValue];
 }
 
+- (void)callJsCallback:(NSTimer *_Nonnull)timer
+{
+  NSArray<JSValue *> *args = [timer userInfo];
```

Comment:
I'm not sure if trying to optimize for the most common case of no additional arguments and setting just the callback as the user info and then checking type here would improve performance or not...